### DISCUSSION TITLE
Error ocurred when building libtorrent using cmake and ninja

### DISCUSSION THREAD

**Discussion by leoncwu** [arvidn/libtorrent#7368] [Q&A]
Hi Arvidn, I am trying to build the library of the latest libtorrent myself. I tried to build with cmake. So, I have cmake, ninja, an boost installed. I was able to get went throught step 1 withour any error. However, when I tried to execute "ninja", the build failed with the following error:
/usr/bin/c++ -DBOOST_ASIO_ENABLE_CANCELIO -DBOOST_ASIO_HAS_STD_CHRONO -DBOOST_ASIO_NO_DEPRECATED -DBOOST_EXCEPTION_DISABLE -DOPENSSL_NO_SSL2 -DTORRENT_BUILDING_LIBRARY -DTORRENT_BUILDING_SHARED -DTORRENT_SSL_PEERS -DTORRENT_USE_LIBCRYPTO -DTORRENT_USE_OPENSSL -D_FILE_OFFSET_BITS=64 -Dtorrent_rasterbar_EXPORTS -I../include -I../deps/try_signal -O3 -DNDEBUG -fPIC -fvisibility=hidden -fvisibility-inlines-hidden -Wall -Wextra -Wpedantic -Wvla -Wno-noexcept-type -Wno-format-zero-length -ftemplate-depth=512 -fexceptions -pthread -std=gnu++14 -MD -MT CMakeFiles/torrent-rasterbar.dir/src/alert_manager.cpp.o -MF CMakeFiles/torrent-rasterbar.dir/src/alert_manager.cpp.o.d -o CMakeFiles/torrent-rasterbar.dir/src/alert_manager.cpp.o -c ../src/alert_manager.cpp
In file included from ../include/libtorrent/torrent_handle.hpp:71:0,
from ../include/libtorrent/alert_types.hpp:50,
from ../include/libtorrent/aux_/alert_manager.hpp:42,
from ../src/alert_manager.cpp:37:
../include/libtorrent/address.hpp:47:40: fatal error: boost/asio/ip/network_v4.hpp: No such file or directory
#include <boost/asio/ip/network_v4.hpp>

Do you happen to know what caused this?

**Comment by arvidn** [comment]
clearly ninja can't find the boost headers. You probably want to look at the cmake output around `find_public_dependency(Boost REQUIRED)` and `target_link_libraries(libtorrent-rasterbar PUBLIC Boost::headers)`

**Comment by leoncwu** [comment]
Hi Arvidn, thank so much for the suggestions. I did some tests, and it seems that the build went through the two functions you mentioned and failed later. Just to show the whole picture, I have to paste the entire message dumped. I am sorry if this post looks awfully long. I am wondering if you would have more comments on how to address the issue.

[0/1] Re-running CMake...
-- debugC done find package: Threads
-- debugC done find package: OpenSSL
-- Found Boost: /usr/include (found version "1.62.0")  
-- debugC done find package: Boost
-- debugC done link library
-- Found Boost: /usr/include (found version "1.62.0") found components: system 
-- debugC done find package: Boost
-- The following features have been enabled:

 * BUILD_SHARED_LIBS, build libtorrent as a shared library
 * dht, enable support for Mainline DHT
 * deprecated-functions, enable deprecated functions for backwards compatibility
 * encryption, Enables encryption in libtorrent
 * exceptions, build with exception support
 * extensions, Enables protocol extensions
 * i2p, build with I2P support
 * logging, build with logging
 * mutable-torrents, Enables mutable torrent support
 * streaming, Enables support for piece deadline

-- The following RECOMMENDED packages have been found:

 * OpenSSL, Full-strength general purpose cryptography library, <https://www.openssl.org/>
   Provides HTTPS support to libtorrent

-- The following REQUIRED packages have been found:

 * Threads
 * Boost

-- The following features have been disabled:

 * static_runtime, build libtorrent with static runtime
 * build_tests, build tests
 * build_examples, build examples
 * build_tools, build tools
 * python-bindings, build python bindings
 * python-egg-info, generate python egg info
 * python-install-system-dir, Install python bindings to the system installation directory rather than the CMake installation prefix
 * gnutls, build using GnuTLS instead of OpenSSL

-- Configuring done
-- Generating done

[1/150] Building CXX object CMakeFiles/torrent-rasterbar.dir/src/disabled_disk_io.cpp.o
FAILED: CMakeFiles/torrent-rasterbar.dir/src/disabled_disk_io.cpp.o 
/usr/bin/c++ -DBOOST_ASIO_ENABLE_CANCELIO -DBOOST_ASIO_HAS_STD_CHRONO -DBOOST_ASIO_NO_DEPRECATED -DBOOST_EXCEPTION_DISABLE -DOPENSSL_NO_SSL2 -DTORRENT_BUILDING_LIBRARY -DTORRENT_BUILDING_SHARED -DTORRENT_SSL_PEERS -DTORRENT_USE_LIBCRYPTO -DTORRENT_USE_OPENSSL -D_FILE_OFFSET_BITS=64 -Dtorrent_rasterbar_EXPORTS -I../include -I../deps/try_signal -O3 -DNDEBUG -fPIC -fvisibility=hidden -fvisibility-inlines-hidden -Wall -Wextra -Wpedantic -Wvla -Wno-noexcept-type -Wno-format-zero-length -ftemplate-depth=512 -fexceptions -pthread -std=gnu++14 -MD -MT CMakeFiles/torrent-rasterbar.dir/src/disabled_disk_io.cpp.o -MF CMakeFiles/torrent-rasterbar.dir/src/disabled_disk_io.cpp.o.d -o CMakeFiles/torrent-rasterbar.dir/src/disabled_disk_io.cpp.o -c ../src/disabled_disk_io.cpp
In file included from ../include/libtorrent/disabled_disk_io.hpp:37:0,
                 from ../src/disabled_disk_io.cpp:35:
../include/libtorrent/io_context.hpp:40:40: fatal error: boost/asio/ts/io_context.hpp: No such file or directory
 #include <boost/asio/ts/io_context.hpp>
                                        ^
compilation terminated.
[2/150] Building CXX object CMakeFiles/torrent-rasterbar.dir/src/disk_io_thread_pool.cpp.o
FAILED: CMakeFiles/torrent-rasterbar.dir/src/disk_io_thread_pool.cpp.o 
/usr/bin/c++ -DBOOST_ASIO_ENABLE_CANCELIO -DBOOST_ASIO_HAS_STD_CHRONO -DBOOST_ASIO_NO_DEPRECATED -DBOOST_EXCEPTION_DISABLE -DOPENSSL_NO_SSL2 -DTORRENT_BUILDING_LIBRARY -DTORRENT_BUILDING_SHARED -DTORRENT_SSL_PEERS -DTORRENT_USE_LIBCRYPTO -DTORRENT_USE_OPENSSL -D_FILE_OFFSET_BITS=64 -Dtorrent_rasterbar_EXPORTS -I../include -I../deps/try_signal -O3 -DNDEBUG -fPIC -fvisibility=hidden -fvisibility-inlines-hidden -Wall -Wextra -Wpedantic -Wvla -Wno-noexcept-type -Wno-format-zero-length -ftemplate-depth=512 -fexceptions -pthread -std=gnu++14 -MD -MT CMakeFiles/torrent-rasterbar.dir/src/disk_io_thread_pool.cpp.o -MF CMakeFiles/torrent-rasterbar.dir/src/disk_io_thread_pool.cpp.o.d -o CMakeFiles/torrent-rasterbar.dir/src/disk_io_thread_pool.cpp.o -c ../src/disk_io_thread_pool.cpp
In file included from ../include/libtorrent/aux_/disk_io_thread_pool.hpp:41:0,
                 from ../src/disk_io_thread_pool.cpp:35:
../include/libtorrent/io_context.hpp:40:40: fatal error: boost/asio/ts/io_context.hpp: No such file or directory
 #include <boost/asio/ts/io_context.hpp>
                                        ^
compilation terminated.
[3/150] Building CXX object CMakeFiles/torrent-rasterbar.dir/src/enum_net.cpp.o
FAILED: CMakeFiles/torrent-rasterbar.dir/src/enum_net.cpp.o 
/usr/bin/c++ -DBOOST_ASIO_ENABLE_CANCELIO -DBOOST_ASIO_HAS_STD_CHRONO -DBOOST_ASIO_NO_DEPRECATED -DBOOST_EXCEPTION_DISABLE -DOPENSSL_NO_SSL2 -DTORRENT_BUILDING_LIBRARY -DTORRENT_BUILDING_SHARED -DTORRENT_SSL_PEERS -DTORRENT_USE_LIBCRYPTO -DTORRENT_USE_OPENSSL -D_FILE_OFFSET_BITS=64 -Dtorrent_rasterbar_EXPORTS -I../include -I../deps/try_signal -O3 -DNDEBUG -fPIC -fvisibility=hidden -fvisibility-inlines-hidden -Wall -Wextra -Wpedantic -Wvla -Wno-noexcept-type -Wno-format-zero-length -ftemplate-depth=512 -fexceptions -pthread -std=gnu++14 -MD -MT CMakeFiles/torrent-rasterbar.dir/src/enum_net.cpp.o -MF CMakeFiles/torrent-rasterbar.dir/src/enum_net.cpp.o.d -o CMakeFiles/torrent-rasterbar.dir/src/enum_net.cpp.o -c ../src/enum_net.cpp
In file included from ../include/libtorrent/enum_net.hpp:50:0,
                 from ../src/enum_net.cpp:40:
../include/libtorrent/io_context.hpp:40:40: fatal error: boost/asio/ts/io_context.hpp: No such file or directory
 #include <boost/asio/ts/io_context.hpp>
                                        ^
compilation terminated.
[4/150] Building CXX object CMakeFiles/torrent-rasterbar.dir/src/mmap_disk_io.cpp.o
FAILED: CMakeFiles/torrent-rasterbar.dir/src/mmap_disk_io.cpp.o 
/usr/bin/c++ -DBOOST_ASIO_ENABLE_CANCELIO -DBOOST_ASIO_HAS_STD_CHRONO -DBOOST_ASIO_NO_DEPRECATED -DBOOST_EXCEPTION_DISABLE -DOPENSSL_NO_SSL2 -DTORRENT_BUILDING_LIBRARY -DTORRENT_BUILDING_SHARED -DTORRENT_SSL_PEERS -DTORRENT_USE_LIBCRYPTO -DTORRENT_USE_OPENSSL -D_FILE_OFFSET_BITS=64 -Dtorrent_rasterbar_EXPORTS -I../include -I../deps/try_signal -O3 -DNDEBUG -fPIC -fvisibility=hidden -fvisibility-inlines-hidden -Wall -Wextra -Wpedantic -Wvla -Wno-noexcept-type -Wno-format-zero-length -ftemplate-depth=512 -fexceptions -pthread -std=gnu++14 -MD -MT CMakeFiles/torrent-rasterbar.dir/src/mmap_disk_io.cpp.o -MF CMakeFiles/torrent-rasterbar.dir/src/mmap_disk_io.cpp.o.d -o CMakeFiles/torrent-rasterbar.dir/src/mmap_disk_io.cpp.o -c ../src/mmap_disk_io.cpp
In file included from ../include/libtorrent/mmap_disk_io.hpp:40:0,
                 from ../src/mmap_disk_io.cpp:42:
../include/libtorrent/io_context.hpp:40:40: fatal error: boost/asio/ts/io_context.hpp: No such file or directory
 #include <boost/asio/ts/io_context.hpp>
                                        ^
compilation terminated.
[5/150] Building CXX object CMakeFiles/torrent-rasterbar.dir/src/choker.cpp.o
FAILED: CMakeFiles/torrent-rasterbar.dir/src/choker.cpp.o 
/usr/bin/c++ -DBOOST_ASIO_ENABLE_CANCELIO -DBOOST_ASIO_HAS_STD_CHRONO -DBOOST_ASIO_NO_DEPRECATED -DBOOST_EXCEPTION_DISABLE -DOPENSSL_NO_SSL2 -DTORRENT_BUILDING_LIBRARY -DTORRENT_BUILDING_SHARED -DTORRENT_SSL_PEERS -DTORRENT_USE_LIBCRYPTO -DTORRENT_USE_OPENSSL -D_FILE_OFFSET_BITS=64 -Dtorrent_rasterbar_EXPORTS -I../include -I../deps/try_signal -O3 -DNDEBUG -fPIC -fvisibility=hidden -fvisibility-inlines-hidden -Wall -Wextra -Wpedantic -Wvla -Wno-noexcept-type -Wno-format-zero-length -ftemplate-depth=512 -fexceptions -pthread -std=gnu++14 -MD -MT CMakeFiles/torrent-rasterbar.dir/src/choker.cpp.o -MF CMakeFiles/torrent-rasterbar.dir/src/choker.cpp.o.d -o CMakeFiles/torrent-rasterbar.dir/src/choker.cpp.o -c ../src/choker.cpp
In file included from ../include/libtorrent/torrent_handle.hpp:71:0,
                 from ../include/libtorrent/alert_types.hpp:50,
                 from ../include/libtorrent/peer_connection_interface.hpp:40,
                 from ../include/libtorrent/peer_connection.hpp:59,
                 from ../src/choker.cpp:38:
../include/libtorrent/address.hpp:47:40: fatal error: boost/asio/ip/network_v4.hpp: No such file or directory
 #include <boost/asio/ip/network_v4.hpp>
                                        ^
compilation terminated.
[6/150] Building CXX object CMakeFiles/torrent-rasterbar.dir/src/disk_job_fence.cpp.o
ninja: build stopped: subcommand failed.

**Comment by userdocs** [comment]
I am just going to guess it's the boost version being `1.62` which is Debian Stretch old.

```
-- Found Boost: /usr/include (found version "1.62.0") found components: system
```

I am pretty sure you need to use a more modern version. You can do this by downloading it, extracting it, and point cmake to this instead.

**Reply by leoncwu** [reply]
Thank you so much for your comments. I want to double-check the Boost library and version. Which version should be good? I saw the following list of available libraries on Debian website. I bet it is reliable to go with stable versions, unless I have to use the latest one.

- [stretch (oldoldstable)](https://packages.debian.org/stretch/libboost-all-dev) (libdevel): Boost C++ Libraries development files (ALL) (default version)
 1.62.0.1: amd64 arm64 armel armhf i386 mips mips64el mipsel ppc64el s390x
- [buster (oldstable)](https://packages.debian.org/buster/libboost-all-dev) (libdevel): Boost C++ Libraries development files (ALL) (default version)
 1.67.0.1: amd64 arm64 armel armhf i386 mips mips64el mipsel ppc64el s390x
- [bullseye (stable)](https://packages.debian.org/bullseye/libboost-all-dev) (libdevel): Boost C++ Libraries development files (ALL) (default version)
 1.74.0.3: amd64 arm64 armel armhf i386 mips64el mipsel ppc64el s390x
- [bookworm (testing)](https://packages.debian.org/bookworm/libboost-all-dev) (libdevel): Boost C++ Libraries development files (ALL) (default version)
 1.74.0.3: amd64 arm64 armel armhf i386 mips64el mipsel ppc64el s390x
- [sid (unstable)](https://packages.debian.org/sid/libboost-all-dev) (libdevel): Boost C++ Libraries development files (ALL) (default version)
 1.74.0.3: alpha amd64 arm64 armel armhf hppa i386 ia64 m68k mips64el mipsel ppc64 ppc64el riscv64 s390x sh4 sparc64 x32
- [experimental](https://packages.debian.org/experimental/libboost-all-dev) (libdevel): Boost C++ Libraries development files (ALL) (default version)
 1.81.0.1~exp1: alpha amd64 arm64 armel armhf hppa i386 ia64 m68k mips64el mipsel ppc64 ppc64el riscv64 s390x sh4 sparc64

**Reply by userdocs** [reply]
I don't use system deps to build boost. I do this.

https://github.com/arvidn/libtorrent/issues/6972#issue-1307236640
